File: playwright-csharp/tests/Reporting/RunArtifactsSupport.cs
using System.Net;
using Reqnroll;
using TestTracks.Playwright.CSharp.Configuration;
using TestTracks.Playwright.CSharp.Specs.Configuration;

namespace TestTracks.Playwright.CSharp.Specs.Reporting;

[Binding]
public static class RunArtifactsSupport
{
    private static readonly Lock Sync = new();
    private static bool _formatterOutputsMoved;
    private static bool _processExitRegistered;

    [BeforeTestRun]
    public static void BeforeTestRun()
    {
        Directory.CreateDirectory(RunArtifacts.Directory);

        if (_processExitRegistered)
        {
            return;
        }

        AppDomain.CurrentDomain.ProcessExit += (_, _) => MoveFormatterOutputs();
        _processExitRegistered = true;
    }

    private static void MoveFormatterOutputs()
    {
        lock (Sync)
        {
            if (_formatterOutputsMoved)
            {
                return;
            }

            MoveIfExists(Path.Combine(RunArtifacts.RootDirectory, "test-tracks-playwright-csharp-report.html"), RunArtifacts.ReportHtmlPath);
            MoveIfExists(Path.Combine(RunArtifacts.RootDirectory, "cucumber-messages.ndjson"), RunArtifacts.CucumberMessagesPath);
            InjectRunDetailsIntoReport();
            DeletePlaywrightScratchArtifacts();
            _formatterOutputsMoved = true;
        }
    }

    private static void MoveIfExists(string sourcePath, string destinationPath)
    {
        if (!File.Exists(sourcePath))
        {
            return;
        }

        Directory.CreateDirectory(Path.GetDirectoryName(destinationPath)!);
        File.Move(sourcePath, destinationPath, overwrite: true);
    }

    private static void InjectRunDetailsIntoReport()
    {
        if (!File.Exists(RunArtifacts.ReportHtmlPath))
        {
            return;
        }

        var html = File.ReadAllText(RunArtifacts.ReportHtmlPath);
        if (html.Contains("id=\"test-tracks-playwright-csharp-run-details\"", StringComparison.Ordinal))
        {
            return;
        }

        var bodyIndex = html.IndexOf("<body>", StringComparison.OrdinalIgnoreCase);
        if (bodyIndex < 0)
        {
            return;
        }

        var insertAt = bodyIndex + "<body>".Length;
        var updated = html.Insert(insertAt, Environment.NewLine + BuildRunDetailsHtml());
        File.WriteAllText(RunArtifacts.ReportHtmlPath, updated);
    }

    private static string BuildRunDetailsHtml()
    {
        var testRun = TestRunConfiguration.Current;
        var settings = testRun.Settings;
        var browser = settings.Browser.ToString().ToLowerInvariant();

        return $$"""
<section id="test-tracks-playwright-csharp-run-details" style="font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif; margin: 16px; padding: 12px 16px; border: 1px solid #d0d7de; border-radius: 6px; background: #f6f8fa; color: #24292f;">
  <strong>Run details</strong>
  <span style="display: inline-block; margin-left: 16px;">Environment: {{Encode(testRun.EnvironmentName)}}</span>
  <span style="display: inline-block; margin-left: 16px;">Browser: {{Encode(browser)}}</span>
  <span style="display: inline-block; margin-left: 16px;">Headless: {{Encode(settings.Headless.ToString().ToLowerInvariant())}}</span>
</section>
""";
    }

    private static string Encode(string value)
    {
        return WebUtility.HtmlEncode(value);
    }

    private static void DeletePlaywrightScratchArtifacts()
    {
        DeleteFiles("*.trace");
        DeleteFiles("*.network");
        DeleteDirectory(Path.Combine(RunArtifacts.Directory, "resources"));
    }

    private static void DeleteFiles(string searchPattern)
    {
        foreach (var path in Directory.EnumerateFiles(RunArtifacts.Directory, searchPattern, SearchOption.TopDirectoryOnly))
        {
            TryDelete(() => File.Delete(path));
        }
    }

    private static void DeleteDirectory(string path)
    {
        if (Directory.Exists(path))
        {
            TryDelete(() => Directory.Delete(path, recursive: true));
        }
    }

    private static void TryDelete(Action action)
    {
        try
        {
            action();
        }
        catch (IOException)
        {
        }
        catch (UnauthorizedAccessException)
        {
        }
    }
}

File: playwright-java/src/test/java/testtracks/reporting/RunArtifactsSupport.java
package testtracks.reporting;

import io.cucumber.java.BeforeAll;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import testtracks.configuration.RunArtifacts;
import testtracks.configuration.TestRunConfiguration;

public final class RunArtifactsSupport {
  private static boolean shutdownHookRegistered;
  private static boolean formatterOutputsMoved;

  private RunArtifactsSupport() {}

  @BeforeAll
  public static void beforeTestRun() throws IOException {
    Files.createDirectories(Path.of(RunArtifacts.directory()));

    if (!shutdownHookRegistered) {
      Runtime.getRuntime().addShutdownHook(new Thread(RunArtifactsSupport::moveFormatterOutputs));
      shutdownHookRegistered = true;
    }
  }

  private static synchronized void moveFormatterOutputs() {
    if (formatterOutputsMoved) {
      return;
    }

    try {
      moveIfExists(
          RunArtifacts.rootDirectory().resolve("test-report.html"), RunArtifacts.reportHtmlPath());
      moveIfExists(
          RunArtifacts.rootDirectory().resolve("cucumber-messages.ndjson"),
          RunArtifacts.cucumberMessagesPath());
      injectRunDetailsIntoReport();
      formatterOutputsMoved = true;
    } catch (Exception ignored) {
    }
  }

  private static void moveIfExists(Path sourcePath, Path destinationPath) throws IOException {
    if (!Files.exists(sourcePath)) {
      return;
    }

    Files.createDirectories(destinationPath.getParent());
    Files.move(sourcePath, destinationPath, java.nio.file.StandardCopyOption.REPLACE_EXISTING);
  }

  private static void injectRunDetailsIntoReport() throws IOException {
    var report = RunArtifacts.reportHtmlPath();
    if (!Files.exists(report)) {
      return;
    }

    var html = Files.readString(report);
    if (html.contains("id=\"test-tracks-playwright-java-run-details\"")) {
      return;
    }

    var bodyIndex = html.toLowerCase().indexOf("<body>");
    if (bodyIndex < 0) {
      return;
    }

    var insertAt = bodyIndex + "<body>".length();
    Files.writeString(
        report,
        html.substring(0, insertAt)
            + System.lineSeparator()
            + runDetailsHtml()
            + html.substring(insertAt));
  }

  private static String runDetailsHtml() {
    var testRun = TestRunConfiguration.current();
    var settings = testRun.settings();
    return
"""
<section id="test-tracks-playwright-java-run-details" style="font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif; margin: 16px; padding: 12px 16px; border: 1px solid #d0d7de; border-radius: 6px; background: #f6f8fa; color: #24292f;">
  <strong>Run details</strong>
  <span style="display: inline-block; margin-left: 16px;">Environment: %s</span>
  <span style="display: inline-block; margin-left: 16px;">Browser: %s</span>
  <span style="display: inline-block; margin-left: 16px;">Headless: %s</span>
</section>
"""
        .formatted(
            escape(testRun.environmentName()),
            escape(settings.browser().name().toLowerCase()),
            escape(Boolean.toString(settings.headless())));
  }

  private static String escape(String value) {
    return value
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace("\"", "&quot;");
  }
}
